Support multiple stacks per network in gird client

0oM4R opened the pull request:
### Description

Using monitoring package, we are getting the available url for services if it is not provided in clien options

### Changes

add `setServiceURLs` that is updating client options with available services urls for only the services that do not associated with url in client options.
this method is used in connect method and before `getDefaultUrl` to insure that we have the available stack url for each service `if it is not provided` 

### Related Issues

- #3078
### Documentation PR

For UI changes, Please provide the Documetation PR on [info_grid](https://github.com/threefoldtech/info_grid)

### Checklist

- [ ] Tests included
- [ ] Build pass
- [ ] Documentation
- [ ] Code format and docstrings
- [ ] Screenshots/Video attached (needed for UI changes)

0oM4R committed:
Support es6 build

0oM4R committed:
Refactor: update events make the error optional and support log with color

0oM4R committed:
- Refactor
  - sendGetRequest to sendRequest and use fetch instead of axios.
  - replace serviceUrl and serviceName with getters
  - add update method that uses URL setter
  - Make all IServiceBase classes take an optional options on their contractor, add error handling for side effect
  -  Grouped all RMB properties to RMB to RMBProps
- Feat:
 - add stats monitor class that monitors stats service
 - add activation monitor class  that monitor activation service

0oM4R committed:
Feat: Add serviceURLManager
it is class managing service URLs with liveness checking <pick the first reachable endpoint for each service>

- Define retries and silent mode settings
- Implement constructor to handle StackManagerOptions
- Add private handleErrorsOnSilentMode method for error handling based on silent mode
- Implement getAvailableStack method to find reachable service URL
- Implement getAvailableServicesStack method to fetch and store available service URLs for all services

0oM4R committed:
Revert: playground/config.js

0oM4R committed:
Refactor: remove sesstion from RMBOptions it will be handled internally

0oM4R committed:
Refactor: update service monitor example

0oM4R committed:
Refactor: Rename service url manager options

0oM4R committed:
Refactor: make tfchainMonitor consotractor takes an optional param

0oM4R committed:
Example: add serviceURlManager Example

0oM4R committed:
Fix: remove error on getting Url before intialization and make it return undefined or emptystring
remove debug lines

0oM4R committed:
Chore: RMB handle error event

0oM4R committed:
Example: add template type to support silent

0oM4R committed:
Chore:
- change naming
- remove error on get url and return empty string in case the url is not set yet

0oM4R committed:
Chore:
- add errror handling in isAlive, will throw in case the service url is not set

0oM4R committed:
chore: export new services

0oM4R committed:
Chore: make tfchain url optional

0oM4R committed:
chore: change stats url to api/summary

0oM4R committed:
chore: add checks to prevent accessing undefined functions in RMB connect and disconnect

0oM4R committed:
Merge branch 'development' of github.com:threefoldtech/tfgrid-sdk-ts into development_monitoring__stacks

0oM4R committed:
chore: validate rmb connectivity though gridproxy health endpoint

0oM4R committed:
chore: add error handling to isAlive

0oM4R committed:
chore: support timeout, and log errors using events

0oM4R committed:
chore: support timeout

0oM4R committed:
refactor: support increasing timeout

0oM4R committed:
fix: return status object instade of throwing error

0oM4R committed:
refactor:
- rename constractor param
- use replace instade of split

0oM4R committed:
Refactor: use /health endpoint in tfchain monitor

0oM4R committed:
refactor: remove logs line

0oM4R committed:
feat: isAlive can take new url

0oM4R committed:
Feat: ping url in parallel

0oM4R committed:
Docs: update example

0oM4R committed:
Refactor: retry logic moved to all stacks, to avoid delay

0oM4R committed:
refactor: enahnce logs

0oM4R committed:
docs: update urls in example

0oM4R committed:
Chore: remove disconnect logic

0oM4R committed:
chore: Remove IDisconnectHandler interface

0oM4R committed:
Chore: update monitoring example

0oM4R committed:
chore: Update the service url on success

0oM4R committed:
Merge branch 'development' of github.com:threefoldtech/tfgrid-sdk-ts into development_monitoring__stacks

0oM4R committed:
feat: Support Service URL manager in grid client

0oM4R committed:
Chore: return if all urls are predefined

0oM4R committed:
Refactor: remove update function and replace it with setter

0oM4R committed:
Refactor: add ServiceBase class

0oM4R committed:
chore: predefine the timeout error

0oM4R committed:
Merge branch 'development_monitoring__stacks' of github.com:threefoldtech/tfgrid-sdk-ts into development_monitoring__stacks_gridclient

0oM4R committed:
merge development

AhmedHanafy725 commented on packages/grid_client/src/client.ts:
```diff
@@ -100,6 +101,14 @@ class GridClient {
+   */
+  async setServiceURLs(): Promise<void> {
+    await getAvailableURLs(this.clientOptions);
+  }
```
what is the added value of this method?

0oM4R committed:
Merge branch 'development' of github.com:threefoldtech/tfgrid-sdk-ts into development_monitoring__stacks_gridclient

MohamedElmdary commented on packages/grid_client/src/manageURLs.ts:
```diff
@@ -0,0 +1,123 @@
+ *
+ * @param {ClientOptions} clientOptions - The client options containing service URLs and network environment.
+ * @returns {Promise<void>} A promise that resolves when the URLs have been retrieved and assigned.
+ */
+export async function getAvailableURLs(clientOptions: ClientOptions) {
+  const { proxyURL, relayURL, substrateURL, graphqlURL, activationURL, network } = clientOptions;
+  const currentURLs = { proxyURL, relayURL, substrateURL, graphqlURL, activationURL };
+  const base = network === NetworkEnv.main ? "grid.tf" : `${network}.grid.tf`;
+  const URLS = {
+    relayURL: [`wss://relay.${base}`, `wss://relay.02.${base}`],
+    proxyURL: [`https://gridproxy.${base}`, `https://gridproxy.02.${base}`],
+    activationURL: [
+      `https://activation.${base}/activation/activate`,
+      `https://activation.02.${base}/activation/activate`,
+    ],
+    graphqlURL: [`https://graphql.${base}/graphql`, `https://graphql.02.${base}/graphql`],
+    substrateURL: [`wss://tfchain.${base}/ws`, `wss://tfchain.02.${base}/ws`],
+  };
+  const missingServicesURLS = getServicesWithoutURLs(currentURLs);
+
+  if (missingServicesURLS.length == 0) return;
+
+  const services = missingServicesURLS.reduce((acc, servicesName) => {
+    acc.push(prepareServices(servicesName, URLS[servicesName]));
+    return acc;
+  }, [] as Service[]);
+  const result = await new ServiceUrlManager({
+    services: services,
+    silent: true,
+  }).getAvailableServicesStack();
+  Object.assign(clientOptions, mapResult(result, missingServicesURLS));
+}
```
I think mutating object byRef might lead to unexpected behaviour specially in vue apps as it turns objects into proxy which rerender ui whenever it gets mutated

So I think we should avoid update objects byref

0oM4R committed:
Merge branch 'development' of github.com:threefoldtech/tfgrid-sdk-ts into development_monitoring__stacks_gridclient

0oM4R committed:
chore:
- make set serviceurls private
- remove mutating clientoptions by ref and return service urls, then update the clientOptions
- make serviceUrlManager work on default mode not silent to throw error
